test/pipeline.test.ts

```diff
@@ -0,0 +1,133 @@
+import {SetupIntegrationTest} from './integration-setup';
+import {v4} from 'uuid';
+
+const {client} = SetupIntegrationTest();
+
+describe('pipelines', () => {
+  it('should be able to run with chaining commands commands', async () => {
+    const key1 = v4();
+    const key2 = v4();
+    const value1 = v4();
+    const value2 = v4();
+    const dictionaryName = v4();
+
+    await client
+      .pipeline()
+      .hset(dictionaryName, key1, value1)
+      .hset(dictionaryName, key2, value2)
+      .set(key1, value1)
+      .set(key2, value2)
+      .exec();
+
+    const results = await client
+      .pipeline()
+      .hget(dictionaryName, key1)
+      .hget(dictionaryName, key2)
+      .get(key1)
+      .get(key2)
+      .exec();
+
+    expect(results).toEqual([
+      [null, value1],
+      [null, value2],
+      [null, value1],
+      [null, value2],
+    ]);
+  });
+  it('should be able to run with batch load commands', async () => {
+    const key1 = v4();
+    const key2 = v4();
+    const value1 = v4();
+    const value2 = v4();
+    const dictionaryName = v4();
+
+    await client
+      .pipeline([
+        ['hset', dictionaryName, key1, value1],
+        ['hset', dictionaryName, key2, value2],
+        ['set', key1, value1],
+        ['set', key2, value2],
+      ])
+      .exec();
+
+    const results = await client
+      .pipeline([
+        ['hget', dictionaryName, key1],
+        ['hget', dictionaryName, key2],
+        ['get', key1],
+        ['get', key2],
+      ])
+      .exec();
+
+    expect(results).toEqual([
+      [null, value1],
+      [null, value2],
+      [null, value1],
+      [null, value2],
+    ]);
+  });
+
+  it('you should be able to mix commands', async () => {
+    const key1 = v4();
+    const key2 = v4();
+    const key3 = v4();
+    const key4 = v4();
+    const value1 = v4();
+    const value2 = v4();
+    const value3 = v4();
+    const value4 = v4();
+    const dictionaryName = v4();
+
+    await client
+      .pipeline([
+        ['hset', dictionaryName, key1, value1],
+        ['hset', dictionaryName, key2, value2],
+        ['set', key3, value3],
+        ['set', key4, value4],
+      ])
+      .hset(dictionaryName, key1, value1)
+      .hset(dictionaryName, key2, value2)
+      .set(key3, value3)
+      .set(key4, value4)
+      .exec();
+
+    const results = await client
+      .pipeline([
+        ['hget', dictionaryName, key1],
+        ['hget', dictionaryName, key2],
+        ['get', key3],
+        ['get', key4],
+      ])
+      .hget(dictionaryName, key1)
+      .hget(dictionaryName, key2)
+      .get(key3)
+      .get(key4)
+      .exec();
+
+    expect(results).toEqual([
+      [null, value1],
```

Comment:
Why are the first entries in the results list expected to be null?

Reply:
I realized this once started running tests against actual Redis in PR that this was what it expected back.

The `exec` return signature is `Promise<[error: Error | null, result: unknown][] | null>`

So was trying to match what was getting back from redis.

Reply:
i had to look this up in the ioredis docs. the return values are `[err, result]`.

Reply:
OK. Let's just an inline comment to the tests explaining this.